### PR TITLE
feat: better stream handling

### PR DESCRIPTION
**Pull request by dignifiedquire**
- upgrade aegir
- upgrade dependencies
- handle stream errors when setting the config
- use pump instead of pipe

I've also pulled in the changes from subcomandante to execa from #89

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -45,19 +43,22 @@
...
-    "run-series": "^1.1.4",
     "shutdown": "^0.2.4",
     "subcomandante": "^1.0.5"
   },
   "devDependencies": {
-    "aegir": "^9.2.1",
+    "aegir": "^9.3.0",
+    "chai": "^3.5.0",
     "mkdirp": "^0.5.1",
+    "multihashes": "^0.3.1",
```
Comment (haadcode) [inline package.json]: Is multihashes needed somewhere else? I don't see it being used in the tests.

**Comment by haadcode** [comment]
One comment and once CI is happy, LGTM.

File: package.json [hunk 2]
```diff
@@ -45,19 +43,20 @@
...
+    "async": "^2.1.4",
+    "execa": "^0.5.0",
     "go-ipfs-dep": "0.4.4",
     "ipfs-api": "^12.0.3",
     "multiaddr": "^2.1.0",
     "once": "^1.4.0",
     "rimraf": "^2.5.4",
-    "run-series": "^1.1.4",
-    "shutdown": "^0.2.4",
-    "subcomandante": "^1.0.5"
```
Comment (diasdavid) [inline package.json]: @jbenet invested a lot of time making sure we had a thing that would not leave zombie IPFS processes in user machines. It was what blocked https://github.com/ipfs/js-ipfsd-ctl/pull/89 all this time.

Until we have comprehensive tests to make sure that you can spawn a bunch of daemons and make sure they don't become zombies, we can't move to execa. Historically, a way to test this was with `station`

If I understand correctly, migrating to `execa` is not a requirement for any particular feature.

Comment (diasdavid) [inline package.json]: As for the tests: https://github.com/ipfs/js-ipfsd-ctl/issues/95

File: src/index.js [hunk 3]
```diff
@@ -5,6 +5,14 @@ const join = require('path').join
...
+  'Addresses.Swarm': ['/ip4/0.0.0.0/tcp/0'],
+  'Addresses.Gateway': '',
+  'Addresses.API': '/ip4/127.0.0.1/tcp/0',
+  disposable: true,
+  init: true
+}
+
 function tempDir () {
   return join(os.tmpdir(), `ipfs_${String(Math.random()).substr(2)}`)
 }
```
Comment (diasdavid) [inline src/index.js]: Is the path to tmpDir available through some call?

File: src/node.js [hunk 4]
```diff
@@ -64,25 +83,20 @@ module.exports = class Node {
     this.initialized = fs.existsSync(path)
     this.clean = true
     this.env = Object.assign({}, process.env, {IPFS_PATH: path})
+    this.disposable = disposable
```
Comment (diasdavid) [inline src/node.js]: add to the list of props `path` as the path used for the repo

File: src/node.js [hunk 5]
```diff
@@ -93,148 +107,108 @@ module.exports = class Node {
...
   // cleanup tmp files
-  // TODO: this is a bad name for a function. a user may call this expecting
-  // something similar to "stopDaemon()". consider changing it. - @jbenet
-  shutdown (done) {
-    if (!this.clean && this.disposable) {
-      rimraf(this.path, (err) => {
-        if (err) throw err
-        done()
-      })
+  shutdown (callback) {
```
Comment (diasdavid) [inline src/node.js]: still a `bad name`, check the deleted comment

Comment (diasdavid) [inline src/node.js]: but independent of the name, it should be document of what it is doing

File: src/node.js (hunk 4 above)
Comment (dignifiedquire) [inline src/node.js]: not sure, I didn't change anything from that logic

File: src/node.js (hunk 5 above)
Comment (dignifiedquire) [inline src/node.js]: I disagree I think it's actuallly a pretty fitting name, in terms of docs nothing changed

File: package.json (hunk 2 above)
Comment (dignifiedquire) [inline package.json]: I don't see how the burden of proof lies on the one changing it, as I haven't seen any proof subcomandante being as good as you say it is. There are tests for the cleanup behavior inside execa, which I wrote when implementing. I also did manual tests with execa inside reusing the testsuite from subcomandante so those cases all work. 

So far all I heard was 'there were extensive tests' but I haven't seen or read what those actually were, so please tell which scenarios you are concerned about and I can add tests for them

File: src/node.js (hunk 4 above)
Comment (dignifiedquire) [inline src/node.js]: this comment was meant for the one above

File: src/node.js (hunk 5 above)
Comment (diasdavid) [inline src/node.js]: There are no docs (am I missing something?)

Comment (dignifiedquire) [inline src/node.js]: There is a comment above which kinda describes what it does, no other docs currently, correct

File: src/node.js (hunk 4 above)
Comment (dignifiedquire) [inline src/node.js]: > add to the list of props path as the path used for the repo

already there if you look at the full file

**Comment by dignifiedquire** [comment]
Added documentation https://ipfs.github.io/js-ipfsd-ctl/

File: package.json (hunk 2 above)
Comment (dignifiedquire) [inline package.json]: @diasdavid I've ported the test suite from node-subcomandante, to run against the new `src/exec.js` to add a baseline of coverage to this module.

**Comment by dignifiedquire** [comment]
Looks like I can't run the last test on travis: 

```
    ✓ SIGTERM kills hang (412ms)
trap: SIGHUP: bad trap
    1) SIGKILL kills survivor
```

File: package.json (hunk 2 above)
Comment (dignifiedquire) [inline package.json]: Soo even though those tests pass fine, it seems thing are wonky right now. I have 5 ipfs instances running after a single `npm test` run :( 

I will revert the switch to execa tomorrow.

**Comment by dignifiedquire** [comment]
@diasdavid this should be ready now, we are back to subcomandante and have more tests and docs

**Pull request merged**